File: src/main/java/me/creepinson/creepinoutils/util/VelocityUtil.java
package me.creepinson.creepinoutils.util;

import dev.throwouterror.util.math.Facing;
import dev.throwouterror.util.math.Tensor;
import net.minecraft.entity.Entity;
import net.minecraft.util.Direction;
import net.minecraft.util.math.MathHelper;

public class VelocityUtil {
    public static final double MIN_SPEED = 0.1;

    public static void accelerate(Entity entity, Facing v) {
        accelerate(entity, v.getDirectionVec());
    }

    public static void accelerate(Entity entity, Direction v) {
        accelerate(entity, Facing.values()[v.ordinal()]);
    }

    public static void accelerate(Entity entity, Tensor v) {
        accelerate(entity, v, MIN_SPEED);
    }

    public static void accelerate(Entity entity, Facing v, double speed) {
        accelerate(entity, v.getDirectionVec(), speed);
    }

    public static void accelerate(Entity entity, Direction v, double speed) {
        accelerate(entity, Facing.values()[v.ordinal()], speed);
    }

    public static void accelerate(Entity entity, Tensor Tensor, double speed) {
        Tensor v = Tensor.clone().mul((float) speed);
        entity.addVelocity(v.x(), v.y(), v.z());
    }

    public static void limitEntitySpeed(Entity entity, double limit) {
        entity.setMotion(MathHelper.clamp(entity.getMotion().x, -limit, limit), MathHelper.clamp(entity.getMotion().y, -limit, limit), MathHelper.clamp(entity.getMotion().z, -limit, limit));
    }

    public static Tensor calculateParabolicVelocity(Tensor from, Tensor to, int heightGain) {
        // Gravity of a potion
        double gravity = 0.115;

        // Block locations
        int endGain = (int) (to.y() - from.y());
        double horizDist = Math.sqrt(distanceSquared(from, to));

        // Height gain

        double maxGain = Math.max(heightGain, (endGain + heightGain));

        // Solve quadratic equation for velocity
        double a = -horizDist * horizDist / (4 * maxGain);
        double b = horizDist;
        double c = -endGain;

        double slope = -b / (2 * a) - Math.sqrt(b * b - 4 * a * c) / (2 * a);

        // Vertical velocity
        double vy = Math.sqrt(maxGain * gravity);

        // Horizontal velocity
        double vh = vy / slope;

        // Calculate horizontal direction
        int dx = (int) (to.x() - from.x());
        int dz = (int) (to.z() - from.z());
        double mag = Math.sqrt(dx * dx + dz * dz);
        double dirx = dx / mag;
        double dirz = dz / mag;

        // Horizontal velocity components
        double vx = vh * dirx;
        double vz = vh * dirz;

        return new Tensor(vx, vy, vz);
    }

    private static double distanceSquared(Tensor from, Tensor to) {
        double dx = to.x() - from.x();
        double dz = to.z() - from.z();

        return dx * dx + dz * dz;
    }
}

File: src/main/kotlin/com/theoparis/creepinoutils/util/VelocityUtil.kt
package com.theoparis.creepinoutils.util

import net.minecraft.entity.Entity
import net.minecraft.util.Direction
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.MathHelper
import net.minecraft.util.math.vector.Vector3d
import net.minecraft.util.math.vector.Vector3i
import kotlin.math.max
import kotlin.math.sqrt

object VelocityUtil {
    const val MIN_SPEED = 0.1

    fun accelerate(entity: Entity, v: BlockPos) {
        accelerate(entity, v.toVector3d(), MIN_SPEED)
    }

    fun accelerate(entity: Entity, v: Direction, speed: Double) {
        accelerate(entity, v.directionVec.toVector3d(), speed)
    }

    fun accelerate(entity: Entity, pos: Vector3d, speed: Double) {
        val v = pos.mul(speed, speed, speed)
        entity.addVelocity(v.x, v.y, v.z)
    }

    fun limitEntitySpeed(entity: Entity, limit: Double) {
        entity.setMotion(
            MathHelper.clamp(entity.motion.x, -limit, limit),
            MathHelper.clamp(entity.motion.y, -limit, limit),
            MathHelper.clamp(entity.motion.z, -limit, limit)
        )
    }

    fun calculateParabolicVelocity(from: Vector3d, to: Vector3d, heightGain: Int): Vector3d {
        // Gravity of a potion
        val gravity = 0.115

        // Block locations
        val endGain = (to.y - from.y).toInt()
        val horizDist = sqrt(distanceSquared(from, to))

        // Height gain
        val maxGain = max(heightGain, endGain + heightGain).toDouble()

        // Solve quadratic equation for velocity
        val a = -horizDist * horizDist / (4 * maxGain)
        val c = -endGain.toDouble()
        val slope = -horizDist / (2 * a) - sqrt(horizDist * horizDist - 4 * a * c) / (2 * a)

        // Vertical velocity
        val vy = sqrt(maxGain * gravity)

        // Horizontal velocity
        val vh = vy / slope

        // Calculate horizontal direction
        val dx = (to.x - from.x).toInt()
        val dz = (to.z - from.z).toInt()
        val mag = sqrt(dx * dx + dz * dz.toDouble())
        val dirx = dx / mag
        val dirz = dz / mag

        // Horizontal velocity components
        val vx = vh * dirx
        val vz = vh * dirz
        return Vector3d(vx, vy, vz)
    }

    private fun distanceSquared(from: Vector3d, to: Vector3d): Double {
        val dx = to.x - from.x
        val dz = to.z - from.z
        return dx * dx + dz * dz
    }
}